numcodecs/jenkins.pyx

```diff
@@ -0,0 +1,327 @@
+"""
+
+import cython
+
+from numcodecs.compat import ensure_contiguous_ndarray
+
+from libc.stdint cimport uint8_t, uint16_t, uint32_t
+
+cpdef uint32_t jenkins_lookup3(const uint8_t[::1] _data, uint32_t initval=0):
+    """
+    jenkins_lookup3(data: bytes, initval: uint32_t=0)
+    hash a variable-length key into a 32-bit value
+
+    data   : the key (unaligned variable-length array of bytes)
+    initval : can be any 4-byte value, defualts to 0
+
+    Returns a 32-bit value.  Every bit of the key affects every bit of
+    the return value.  Two keys differing by one or two bits will have
+    totally different hash values.
+
+    The best hash table sizes are powers of 2.  There is no need to do
+    mod a prime (mod is sooo slow!).  If you need less than 32 bits,
+    use a bitmask.  For example, if you need only 10 bits, do
+    h = (h & hashmask(10))
+    In which case, the hash table should have hashsize(10) elements.
+
+    If you are hashing strings, do it like this:
+    ```
+    h = 0
+    for k in strings:
+      h = jenkins_lookup3(k, h)
+    ```
+
+    By Bob Jenkins, 2006.  bob_jenkins@burtleburtle.net.  You may use this
+    code any way you wish, private, educational, or commercial.  It's free.
+
+    Use for hash table lookup, or anything where one collision in 2^^32 is
+    acceptable.  Do NOT use for cryptographic purposes.
+
+    Converted from H5_checksum_lookup3
+    https://github.com/HDFGroup/hdf5/blob/577c192518598c7e2945683655feffcdbdf5a91b/src/H5checksum.c#L378-L472
+
+    Originally hashlittle from https://www.burtleburtle.net/bob/c/lookup3.c
+    Alternatively, consider the hashword implementation if we can assume little endian and alignment.
+    """
+
+    cdef:
+        size_t length = _data.shape[0]
+        # internal state
+        uint32_t a, b, c = 0
+
+    # Set up the internal state
+    a = b = c = (<uint32_t>0xdeadbeef) + (<uint32_t>length) + initval
+
+    # Return immediately for empty bytes
+    if length == 0:
+        return c
+
+    cdef:
+        const uint8_t *k = <const uint8_t *> &_data[0]
+
+    # We are adding uint32_t values (words) byte by byte so we do not assume endianness or alignment
+    # lookup3.c hashlittle checks for alignment
+
+    # all but the last block: affect some 32 bits of (a,b,c)
+    while length > 12:
+        a += k[0]
+        a += (<uint32_t>k[1]) << 8
+        a += (<uint32_t>k[2]) << 16
+        a += (<uint32_t>k[3]) << 24
+        b += k[4]
+        b += (<uint32_t>k[5]) << 8
+        b += (<uint32_t>k[6]) << 16
+        b += (<uint32_t>k[7]) << 24
+        c += k[8]
+        c += (<uint32_t>k[9]) << 8
+        c += (<uint32_t>k[10]) << 16
+        c += (<uint32_t>k[11]) << 24
+        a, b, c = _jenkins_lookup3_mix(a, b, c)
+        length -= 12
+        k += 12
+
+    # -------------------------------- last block: affect all 32 bits of (c)
+    if length == 12:
+        c += (<uint32_t>k[11]) << 24
+        length -= 1
+        
+    if length == 11:
```

Comment:
Should these be `elif`?

It's worth noting that Cython can detect cases like these when they are written as `if...elif...else` and convert them to `switch...case` in C/C++